tests/handlers/test_appservice.py

```diff
@@ -849,6 +861,125 @@ def test_application_services_receive_bursts_of_to_device(self) -> None:
+                        self.exclusive_as_user_3_device_id: message_content
+                    },
+                }
+            },
+            access_token=self.local_user_token,
+        )
+        self.assertEqual(chan.code, 200, chan.result)
+
+        # Have exclusive_as_user send a to-device message to local_user
+        for user_token in [
+            self.exclusive_as_user_token,
+            self.exclusive_as_user_2_token,
+            self.exclusive_as_user_3_token,
+        ]:
+            chan = self.make_request(
+                "PUT",
+                "/_matrix/client/r0/sendToDevice/m.room_key_request/4",
+                content={
+                    "messages": {
+                        self.local_user: {self.local_user_device_id: message_content}
+                    }
+                },
+                access_token=user_token,
+            )
+            self.assertEqual(chan.code, 200, chan.result)
+
+        # Check if our application service - that is interested in exclusive_as_user - received
+        # the to-device message as part of an AS transaction.
+        # Only the local_user -> exclusive_as_user to-device message should have been forwarded to the AS.
+        #
+        # The uninterested application service should not have been notified at all.
+        self.send_mock.assert_called_once()
+        (
+            service,
+            _events,
+            _ephemeral,
+            to_device_messages,
+            _otks,
+            _fbks,
+            _device_list_summary,
+        ) = self.send_mock.call_args[0]
+
+        # Assert that this was the same to-device message that local_user sent
+        self.assertEqual(service, interested_appservice)
+
+        # Assert expected number of messages
+        self.assertEqual(len(to_device_messages), 3)
+
+        device_msg_1 = {
+            "msg": to_device_messages[0],
+            "expected_user": self.exclusive_as_user,
+            "expected_device": self.exclusive_as_user_device_id,
+        }
+
+        device_msg_2 = {
+            "msg": to_device_messages[1],
+            "expected_user": self.exclusive_as_user_2,
+            "expected_device": self.exclusive_as_user_2_device_id,
+        }
+
+        device_msg_3 = {
+            "msg": to_device_messages[2],
+            "expected_user": self.exclusive_as_user_3,
+            "expected_device": self.exclusive_as_user_3_device_id,
+        }
+
+        for device_msg in [device_msg_1, device_msg_2, device_msg_3]:
+            self.assertEqual(device_msg["msg"]["type"], "m.room_key_request")
+            self.assertEqual(device_msg["msg"]["sender"], self.local_user)
+            self.assertEqual(
+                device_msg["msg"]["to_user_id"], device_msg["expected_user"]
+            )
+            self.assertEqual(
+                device_msg["msg"]["to_device_id"], device_msg["expected_device"]
+            )
+            self.assertEqual(device_msg["msg"]["content"], message_content)
```

Comment:
I think adding the in-between mapping is more confusing than just doing the asserts?

```suggestion
        for device_msg in to_device_messages:
            self.assertEqual(device_msg["type"], "m.room_key_request")
            self.assertEqual(device_msg["sender"], self.local_user)
            self.assertEqual(device_msg["content"], message_content)

        self.assertEqual(
            to_device_messages[0]["to_user_id"], self.exclusive_as_user
        )
        self.assertEqual(
            to_device_messages[0]["to_device_id"], self.exclusive_as_user_device_id,
        )

        self.assertEqual(
            to_device_messages[1]["to_user_id"], self.exclusive_as_user_2
        )
        self.assertEqual(
            to_device_messages[1]["to_device_id"], self.exclusive_as_user_2_device_id,
        )

        self.assertEqual(
            to_device_messages[2]["to_user_id"], self.exclusive_as_user_3
        )
        self.assertEqual(
            to_device_messages[2]["to_device_id"], self.exclusive_as_user_3_device_id,
        )
```